### ISSUE TITLE
NativeTestSuite does not expose binaries

### ISSUE THREAD

**Issue by pedrolamarao** [nokeedev/gradle-native#574]
Currently, the `NativeTestSuite` interface does not expose a `binaries` property. 
(This is according with published Javadoc and tested with most recent 0.4.x snapshot)
This is not a problem for Groovy scripts that dynamically resolve things.
But in Kotlin one is forced to do things like this:

```
testSuites {
    register("test", NativeTestSuite::class.java).get().apply {
        testedComponent(library)

        dependencies {
            implementation(project(":googletest"))
        }

        if (this is DefaultNativeTestSuiteComponent) {
            binaries.configureEach {
                if (this is ExecutableBinary) {
                    compileTasks.configureEach {
                        compilerArgs.addAll("-std=c++20", "-flto")
                    }
                    linkTask {
                        // #XXX: Nokee ignores toolchain linker arguments
                        linkerArgs.addAll("-fuse-ld=lld", "-flto")
                    }
                }
            }
        }
    }
}
```

**Comment by lacasseio** [comment]
You are absolute right, those are the current things I'm considering fixing while I pass by :-) There's a higher cross-cutting concern that I'm still debating how to fix regarding static typing which revolves around how interface should be implemented by components/variants/binaries/etc. We have some _special_ interfaces that are used to _group_ domain object in a Gradle collection-like OO, i.e. `BinaryAwareComponent`, `SourceAwareComponent`, etc. Those special interface signals that a component is aware of a specific domain object type, i.e. `Binary`, `LanguageSourceSet`, etc. Users can then _view_ the domain object and configure them however they want.

It's pretty straightforward for users to understand, however, it increases the configuration complexity. Let me explain. In your example, you configure the `ExecutableBinary` using `binaries.configureEach`. which is not exactly the correct intention. In Kotlin DSL, you caught one issue quickly thanks to the type-safety nature of the language. The correct intention is to configure the type `ExecutableBinary` which should be `binaries.configureEach(ExecutableBinary)`. This brings us to the second issue with those `*AwareComponent`, the intention is to configure **the** executable binary as opposed to all binaries fulfilling the `ExecutableBinary` contract.

This leads to the composition of domain objects using various specialized interfaces most specifically type-safety. A generic native test suite could be different than a Google Test test suite which can also be different from than CppUnit test suite. Most of them should use an executable binary as their entry point but it may not be strictly necessary. The issue here is to avoid being too opinionated in the higher type causing lower types (individual test suite type) to have unnecessary properties. The universal model underneath allows for runtime composition but at the DSL level, it's a bit complicated to statically _switch_ between interfaces. The leading solution is to use/abuse the `ExtensionAware` interface which has special handling by Gradle to generate Kotlin extension methods thus allowing composition without requiring a _special_ mechanic for selecting the right implementation interface.

Long story short, there was a reason why we didn't implement all `*AwareComponent` interfaces on `NativeTestSuite` directly. We getting close to a general solution for all Nokee plugins which is what we will use here. I will leave the issue open as it's valid and users should expect to have access to the binaries.
